File: 0112-path-sum/0112-path-sum.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  bool hasPathSum(TreeNode? root, int targetSum) {    
    bool has_sum(TreeNode? root, int currentSum) {
      if (root == null) {
        return false;
      }

      currentSum += root.val;

      // if we reach a leaf we want to check targetSum
      if (root.right == null && root.left == null) {
        return currentSum == targetSum;
      }

      return has_sum(root.left, currentSum) || has_sum(root.right, currentSum);
    }
  
    // the local variable currentSum initial value is 0
    return has_sum(root, 0);
  }
}

File: 0112-path-sum/0112-path-sum.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# in this approach we do not need to track the sum 
# ech time we march forward in the tree we sub the value 
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root : return False
        
        if not root.left and not root.right:
            return root.val == targetSum
        
        return self.hasPathSum(root.left, targetSum - root.val) or self.hasPathSum(root.right, targetSum - root.val)
    
# print(Solution().hasPathSum(TreeNode(1, TreeNode(2), TreeNode(3)), 3))